test/test_nn.py

```diff
@@ -7699,6 +7701,39 @@ def test_adaptive_log_softmax(self):
         out = asfm.predict(x)
         self.assertEqual(out, asfm.log_prob(x).argmax(dim=1))
 
+    def test_transformer_number_match(self):
+        # Train the simple copy task.
+        V = 11
+        num_encoder_layers = 1
+        num_decoder_layers = 1
+        d_model = 16  # embedding
+        nhead = 2  # number of heads
+        dropout = 0.20
+        d_ff = 64
+        batch_size = 400
+        milliseconds = int(round(time.time() * 1000))
+        torch.manual_seed(milliseconds)
+
+        model = Transformer(V, V, d_model=d_model, nhead=nhead, num_encoder_layers=num_encoder_layers, 
+                            num_decoder_layers=num_decoder_layers, d_ff=d_ff)
+        model_opt = NoamOpt(model.encoder.src_embed.d_model, 1, 400,
+                            torch.optim.Adam(model.parameters(), lr=2.5, betas=(0.9, 0.98), eps=1e-9))
+        criterion = LabelSmoothingLoss(size=V, padding_idx=0, smoothing=0.0)
+
+        print("Training the model...")
+        for epoch in range(10):
+            model.train()
+            train_epoch(data_gen(V, batch_size, 21), model, criterion, model_opt) 
```

Comment:
I'd move the Embedding outside of the Transformer. That is, the result of data_gen will be passed through an instance of nn.Embedding outside of this forward call. The reasoning for that is, that someone might want to use a different kind of Embedding and will then need to reimplement their own Transformer. We also don't do this for LSTM.

Reply:
This is crucial for libs like allennlp or pytext for example.